dismiss service for persistent notifications

tedstriker opened the pull request:
## Description:
Unnecessary notifications can now be removed automatically. Added a
dismiss service to remove persistent notifications via script and/or
automation.

**Related issue (if applicable):** fixes #<home-assistant issue number goes here>
https://community.home-assistant.io/t/programmatically-dismissing-a-persistent-notification/10968

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#2693

## Example entry for `configuration.yaml` (if applicable):
```yaml
it rather concerns scripts and automations
```

## Checklist:

If user exposed functionality or configuration variables are added/changed:
  - [X] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)
https://github.com/home-assistant/home-assistant.github.io/pull/2693

If the code communicates with devices, web services, or third-party tools:
  - [X] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [X] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [X] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [X] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [X] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [X] Tests have been added to verify that the new code works.

[ex-requir]:
[ex-import]:

tedstriker committed:
dismiss service for persistent notifications

Unnecessary notifications can now be removed automatically. Added a
dismiss service to remove persistent notifications via script and/or
automation.

tedstriker committed:
Merge remote-tracking branch 'home-assistant/dev' into dev

tedstriker committed:
fixed test

added missing pause

homeassistant commented:
Hi @tedstriker, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#7727).

Once you do that we will be able to review and accept this pull request.

Thanks!

tedstriker commented:
I'm quite unexperienced with pull requests... can somebody tell me were it got wrong?
Tox ran through without problems on my system (I only have Python 3.6 installed)
Also I don't know where this `Merge remote-tracking branch 'home-assistant/dev' into dev` commit is coming from. Is there a way to remove it?

balloob commented on homeassistant/components/persistent_notification.py:
```diff
@@ -92,12 +114,29 @@ def create_service(call):
+        if notification_id is not None:
+            entity_id = ENTITY_ID_FORMAT.format(slugify(notification_id))
+        else:
+            entity_id = async_generate_entity_id(
```
Don't use generate entity id because that will always generate a unique one.

balloob replied:
Actually, notification id is required in the schema so this branch will never be hit.

tedstriker replied:
has been changed as well

balloob commented on homeassistant/components/persistent_notification.py:
```diff
@@ -57,6 +67,18 @@ def async_create(hass, message, title=None, notification_id=None):
 
 
+@callback
+def async_dismiss(hass, notification_id=None):
```
notification_id should be required as per the schema.

balloob commented on homeassistant/components/persistent_notification.py:
```diff
@@ -57,6 +67,18 @@ def async_create(hass, message, title=None, notification_id=None):
+@callback
+def async_dismiss(hass, notification_id=None):
+    """Remove a notification."""
+    data = {
```
No need for this whole construct as notification_id cannot be None.

tedstriker replied:
has been changed

balloob commented on homeassistant/components/persistent_notification.py:
```diff
@@ -43,6 +48,11 @@ def create(hass, message, title=None, notification_id=None):
     hass.add_job(async_create, hass, message, title, notification_id)
 
 
+def dismiss(hass, notification_id=None):
```
Notification id is required.

balloob commented on tests/components/test_persistent_notification.py:
```diff
@@ -64,3 +64,16 @@ def test_create_template_error(self):
+        """Ensure removal of specific notification"""
+        assert len(self.hass.states.entity_ids(pn.DOMAIN)) == 0
+
+        pn.create(self.hass, 'test', notification_id='Beer 2')
```
🍻

tedstriker committed:
Removed unnecessary else branches and inits

The unnecessary branches and fallback inits have been removed. Hope
it’s acceptable now.

tedstriker committed:
added missing docstring

new function was lacking a description

tedstriker committed:
added missing period in comments (D400)

also changed order of import … and from … as error I100 was raised.

tedstriker committed:
Resolve conflict after PR #7658

Changed run_in_executor to async_add_job

tedstriker reviewed (commented):
changes have been made as demanded

tedstriker committed:
Merge remote-tracking branch 'home-assistant/dev' into dev

balloob commented:
Your branch seems to have been broken, 134 files changed. Please repair

tedstriker committed:
update changes

update fork

tedstriker commented:
As I don't know how to merge in the recent updates into my fork to fix the issues, I'll just delete it and create a new pull request.

tedstriker commented:
new PR: https://github.com/home-assistant/home-assistant/pull/7996
